### PR TITLE
feat: Prompt suggestions for ui.chatbot #2250

### LINKED ISSUE #2250: Buttons as an input format for the chatbot widget
### Buttons as an input format for the chatbot widget

I propose adding a feature that allows users to customize the output of the ui.chatbot widget into a form with interactive buttons.
This would enhance the user experience by providing a more interactive and accessible way to navigate the chatbot's responses.

image

### Problem it Solves
Currently, the chatbot widget allows a [feedback interaction](https://github.com/h2oai/wave/issues/2165), but this new feature would be a generalization by having a customizable form with any buttons and options.
Sometimes, I would like the input from the user to be a closed answer within limited possibilities.

### Suggested Implementation
- Allow users to define the structure of the form and the labels on the buttons.
- Whenever the form is display as input option, the regular text input would de disabled

### Potential Benefits
- Enhances user engagement with a more interactive interface.
- Simplifies user interaction, especially for complex queries or multiple-choice questions.
- Makes the chatbot more accessible, particularly for users who prefer a guided interaction.

### PR DESCRIPTION
**Pull request by marek-mihok**
**The PR fulfills these requirements:** (check all the apply)

- [x] It's submitted to the `main` branch.
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `feat: Add a button #xxx`, where "xxx" is the issue number).
- [x] When resolving a specific issue, the PR description includes `Closes #xxx`, where "xxx" is the issue number.
- [x] If changes were made to `ui` folder, unit tests (`make test`) still pass.
- [x] New/updated tests are included

___

This PR brings `suggestions: ChatSuggestion[]` attribute to show clickable prompt suggestions under the last response in ui.chatbot component. When the suggestion is clicked, it fires `suggestion` event which can subsequently be handled by the user.

The user input can be limited to prompt suggestions only by specifying `disabled` prop which disables user input or it can be used in a combination with text input as well.

chatbot-prompt-suggestions

Updated API is as follows:
```tsx
/** Create a chat prompt suggestion displayed as button below the last response in the chatbot component. */
export interface ChatSuggestion {
  /** An identifying name for this component. */
  name: Id
  /** The text displayed for this suggestion. */
  label: S
  /** The caption displayed below the label. */
  caption?: S
  /** The icon to be displayed for this suggestion. */
  icon?: S
}

/** Create a chatbot card to allow getting prompts from users and providing them with LLM generated answers. */
export interface Chatbot {
  /** An identifying name for this component. */
  name: Id
  /** Chat messages data. Requires list or cyclic buffer. */
  data: Rec
  /** Chat input box placeholder. Use for prompt examples. */
  placeholder?: S
  /** The events to capture on this chatbot. One of 'stop' | 'scroll_up' | 'feedback' | 'prompt_suggestion'. */
  events?: S[]
  /** True to show a button to stop the text generation. Defaults to False. */
  generating?: B
  /** The previous messages to show as the user scrolls up. */
  prev_items?: ChatbotMessage[]
  /** Clickable prompt suggestions shown below the last response. */
  suggestions?: ChatSuggestion[]
  /** True if the user input should be disabled. */
  disabled?: B
}
```

The same API update also applies to `ui.chatbot_card` component.

Closes #2250

### CODE REVIEW

**Comment by marek-mihok** [comment]
First 2 points are done:

- Initial "head start" suggestions with captions and icons

https://github.com/h2oai/wave/assets/23740173/feabd35c-41c3-40d6-9fd3-781e1860b749

- Prompt "in-chat" suggestions

https://github.com/h2oai/wave/assets/23740173/826238cb-bd9a-4fd6-a598-ea20212b885f


> The example could use some polish as well, it's very rough at the moment.

@mturoci what do you mean by rough? In terms of code quality or the use case relevance? Can you give me some thoughts about what you consider a proper example?

**Comment by mturoci** [comment]
> what do you mean by rough?

Would that demo persuade you to use Wave? Probably not. Would be better to append the suggestion right away, but add a slight timeout on suggestion selection to emulate computation, hide suggestions in the meantime, then append the response.

**Comment by mturoci** [comment]
> First 2 points are done:

Looks better, the border radius is off though. Try to use the same as used in the rest of our cards/components.

File: ui/src/chatbot.tsx [hunk 1]
```diff
@@ -115,6 +164,7 @@ export const XChatbot = (props: Chatbot) => {
     botTextColor = React.useMemo(() => getContrast(theme.palette.neutralLighter), [theme.palette.neutralLighter]),
     msgContainerRef = React.useRef<HTMLDivElement>(null),
     skipNextBottomScroll = React.useRef(false),
+    hasSomeCaption = React.useMemo(() => props.prompt_suggestions?.some(({ caption }) => !!caption), [props.prompt_suggestions]),
```
Comment (mturoci) [inline ui/src/chatbot.tsx:167]: No need for memo here. We expect a few suggestions only so safe to recompute on every render.

File: ui/src/chatbot.tsx [hunk 2]
```diff
@@ -212,6 +265,21 @@ export const XChatbot = (props: Chatbot) => {
             </span>
           </div>
         ))}
+        {props.prompt_suggestions && props.prompt_suggestions.length > 0 &&
```
Comment (mturoci) [inline ui/src/chatbot.tsx:268]: nit

```suggestion
        {!!props.prompt_suggestions?.length &&
```

File: website/docs/examples/assets/chatbot-events-suggestions.png
Comment (mturoci) [inline website/docs/examples/assets/chatbot-events-suggestions.png:1]: Update.

File: ui/src/chatbot.tsx [hunk 3]
```diff
@@ -264,10 +333,14 @@ interface State {
   data: Rec
   /** Chat input box placeholder. Use for prompt examples. */
   placeholder?: S
-  /** The events to capture on this chatbot. One of 'stop' | 'scroll_up' | 'feedback'. */
+  /** The events to capture on this chatbot. One of 'stop' | 'scroll_up' | 'feedback' | 'prompt_suggestion'. */
   events?: S[]
   /** True to show a button to stop the text generation. Defaults to False. */
   generating?: B
+  /** Clickable prompt suggestions shown below the last response. */
+  prompt_suggestions?: ChatPromptSuggestion[]
```
Comment (mturoci) [inline ui/src/chatbot.tsx:341]:
```suggestion
  suggestions?: ChatSuggestion[]
```

**Review by mturoci — CHANGES_REQUESTED** [review]
Looks good, just a few small comments.

**Review by mturoci — APPROVED** [review]

**Pull request merged**